src/ring_theory/polynomial/vieta.lean

```diff
@@ -0,0 +1,153 @@
+
+lemma coeff_top (n : ℕ) (r : ℕ → α) : coeff (∏ i in range n, (X + C (r i))) n = 1 :=
+begin
+  convert (monic_prod_X_add_C n r).leading_coeff,
+  exact (nat_degree_prod_X_add_C n r).symm,
+end
+
+lemma sum_prod_mul_X_add_C_aux (k n : ℕ) (r : ℕ → α) : ∑ x in range k,
+  (∑ A in powerset_len (n - x) (range n),
+    ∏ j in A, r j) * (X + C (r n)).coeff (k + 1 - x) = 0 :=
+begin
+  rw [← nsmul_zero (range k).card, ← sum_const],
+  refine sum_congr rfl (λ x hx, _),
+  rw [coeff_ge_two_X_add_C, mul_zero],
+  rw add_comm k _,
+  exact nat.lt_sub_right_of_add_lt (add_lt_add_left (mem_range.mp hx) 1),
+end
+
+lemma sum_prod_mul_X_add_C (k n : ℕ) (h : k ≤ n) : ∑ l in range (k + 1),
+  (∑ A in powerset_len (n - l) (range n), ∏ j in A, r j) * (X + C (r n)).coeff (k - l) =
+  (∑ A in powerset_len (n - k + 1) (range n), ∏ j in A, r j)
+    + (∑ A in powerset_len (n - k) (range n), ∏ j in A, r j) * (r n) :=
+begin
+  rw [sum_range_succ, nat.sub_self, coeff_zero_X_add_C],
+  cases k with k, { simp, },
+  rw [nat.succ_eq_add_one, sum_range_succ, nat.add_sub_cancel_left, coeff_one_X_add_C, mul_one,
+    sum_prod_mul_X_add_C_aux, add_zero, add_comm],
+  congr,
+  omega nat,
+end
+
+lemma sum_prod_insert {n k : ℕ} :
+  ∑ A in powerset_len (n - k) (range n), (∏ j in A, r j) * r n =
+    ∑ A in powerset_len (n - k) (range n), ∏ j in insert n A, r j :=
+begin
+  refine sum_congr rfl (λ t ht, _),
+  rw [mul_comm, prod_insert],
+  intro hn,
+  apply @not_mem_range_self n,
+  exact mem_of_subset (mem_powerset_len.mp ht).left hn,
+end
+
+lemma powerset_len_sum_succ_insert_aux {n k : ℕ} (h : k ≤ n) :
+  ∑ a in filter (λ (a : finset ℕ), (insert n a).card = n - k + 1) (range n).powerset,
+    ∏ j in insert n a, r j =
+  ∑ a in filter (λ (a : finset ℕ), a.card = n - k) (range n).powerset,
+    ∏ j in insert n a, r j :=
+begin
+  refine sum_congr _ (λ _ _, rfl),
+  ext,
+  rw [mem_filter, mem_filter, and.congr_right_iff],
+  intro ha,
+  rw card_insert_of_not_mem,
+  { exact add_left_inj _, },
+  rw mem_powerset at ha,
+  intro h,
+  apply not_mem_range_self,
+  exact mem_of_subset ha h
+end
+
+lemma powerset_len_sum_succ_insert (n k : ℕ) (h : k ≤ n) :
+  ∑ A in powerset_len (n - k + 1) (range n), ∏ j in A, r j +
+  ∑ A in powerset_len (n - k) (range n), ∏ j in insert n A, r j =
+  ∑ A in powerset_len (n - k + 1) (range (n + 1)), ∏ j in A, r j :=
+begin
+  rw [@powerset_len_eq_filter _ _ (range (n + 1)), range_add_one, sum_filter, sum_powerset_insert,
+    ← sum_filter, ← sum_filter, ← powerset_len_eq_filter, powerset_len_sum_succ_insert_aux h,
+    ← @powerset_len_eq_filter],
+  exact not_mem_range_self,
+end
+
+/-- A sum version of Vieta's Formulas which shows that the `k`th coefficient of a product
+of linear terms `X + r i` is the `k`th symmetric polynomial of the constant terms `r i`. -/
+lemma coeff_of_prod_X_add_C :
+  ∀ {k : ℕ}, k ≤ n → coeff (∏ i in range n, (X + C (r i))) k
+  = ∑ A in (powerset_len (n - k) (range n)), (∏ j in A, r j) :=
```

Comment:
I think we should explicitly define the `k`th symmetric polynomial, and use that definition here.
#3169 contains a draft of general symmetric polynomials. Feel free to use or modify parts of it. If you think a definition in terms of `powerset_len` is easiest to work with, go ahead.
This probably means that you should put this Vieta PR on hold, and make a small PR about symmetric polynomials first. Putting everything in 1 big PR will make the reviewing difficult.

What do you think?

Reply:
Sure, I'll take #3169 up to the definition for elementary polynomials. Your file looks like it's mostly complete, anyway, so it shouldn't take too much time to make an isolated PR out of it.

How do I put this PR on hold?

Reply:
We mostly pay attention to the PRs labeled "awaiting-review" so don't worry about leaving this one alone for a bit, as long as you're intending to come back to it eventually. You could also add a "WIP" label.

If you're planning to work directly on #3169, you can mark this PR as dependent on that one by adding `- [ ] depends on: #3169` to the bottom of your PR comment above. (Similarly if you're planning to open a new PR.)

Reply:
@acxxa If you want to work on #3169 that's great! But I think some of the theorems on symmetric polynomials are only half done. So you might want to split the PR into pieces...
(Feel free to do whatever you want with the code... I don't consider it "mine".)